packages/SwingSet/src/liveslots/watchedPromises.js

```diff
@@ -0,0 +1,186 @@
+// @ts-check
+
+// no-lonely-if is a stupid rule that really should be disabled globally
+/* eslint-disable no-lonely-if */
+
+import { assert } from '@agoric/assert';
+import { M } from '@agoric/store';
+import { parseVatSlot } from '../lib/parseVatSlots.js';
+
+export function makeWatchedPromiseManager(
+  syscall,
+  vrm,
+  vom,
+  cm,
+  convertValToSlot,
+  convertSlotToVal,
+) {
+  const { makeScalarBigMapStore } = cm;
+  const { defineDurableKind } = vom;
+
+  // watched promises by vpid: each entry is an array of watches on the
+  // corresponding vpid; each of these is in turn an array of a watcher object
+  // and the arguments associated with it by `watchPromise`.
+  let watchedPromiseTable;
+
+  function pseudoThen(p, vpid) {
+    function settle(value, wasFulfilled) {
+      const watches = watchedPromiseTable.get(vpid);
+      watchedPromiseTable.delete(vpid);
+      for (const watch of watches) {
+        const [watcher, ...args] = watch;
+        Promise.resolve().then(() => {
+          if (wasFulfilled) {
+            if (watcher.onFulfilled) {
+              watcher.onFulfilled(value, ...args);
+            }
+          } else {
+            if (watcher.onRejected) {
+              watcher.onRejected(value, ...args);
+            } else {
+              throw value;
+            }
+          }
+        });
+      }
+    }
+
+    p.then(
+      res => settle(res, true),
+      rej => settle(rej, false),
+    );
+  }
+
+  function loadWatchedPromiseTable() {
+    const deadPromisesRaw = syscall.vatstoreGet('deadPromises');
+    let deadPromises;
+    if (deadPromisesRaw) {
+      deadPromises = new Set(deadPromisesRaw.split(','));
+      syscall.vatstoreDelete('deadPromises');
+    } else {
+      deadPromises = new Set();
+    }
+    for (const [vpid, watches] of watchedPromiseTable.entries()) {
+      if (deadPromises.has(vpid)) {
+        watchedPromiseTable.delete(vpid);
+        for (const watch of watches) {
+          const [watcher, ...args] = watch;
+          if (watcher.onRejected) {
+            Promise.resolve().then(() => {
+              watcher.onRejected('vat upgraded', ...args);
+            });
+          }
```

Comment:
This was surprisingly inconsistent with the normal rejection behavior which throws the error if `onRejected` doesn't exist, but I suppose we wouldn't win much from throwing a `'vat upgraded'` for very single of these now dead promises.

In a future where `watchPromise` creates a chained virtual promise, we may be able to do better, but maybe not.

Reply:
I think this was just an oversight on my part.  I'll tweak.